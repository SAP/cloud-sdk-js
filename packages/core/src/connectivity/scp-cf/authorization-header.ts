import {
  unixEOL,
  isNullish,
  createLogger,
  pickIgnoreCase,
  toSanitizedObject,
  encodeBase64
} from '@sap-cloud-sdk/util';
import type { ODataRequest, ODataRequestConfig } from '../../odata-common';
import {
  AuthenticationType,
  Destination,
  DestinationAuthToken,
  sanitizeDestination
} from './destination';

const logger = createLogger({
  package: 'core',
  messageContext: 'authorization-header'
});

/**
 * @deprecated Since v1.20.0. Use [[buildAuthorizationHeaders]] instead.
 * Adds authorization headers for a given ODataRequest to existing headers.
 *
 * @param request - an ODataRequest.
 * @param headers - The headers that should be added to.
 * @returns The provided headers with the new authorization headers.
 */
export async function addAuthorizationHeader<
  RequestT extends ODataRequestConfig
>(
  request: ODataRequest<RequestT>,
  headers: Record<string, string>
): Promise<Record<string, string>> {
  const destination = request.destination;
  if (!destination) {
    return headers;
  }
  const authHeaders = await getAuthHeaders(
    destination,
    request.config.customHeaders
  );
  return {
    ...headers,
    ...authHeaders
  };
}

function needsAuthHeaders(destination: Destination): boolean {
  if (!destination.authentication) {
    return false;
  }
  const authTypesWithAuthorizationHeader: AuthenticationType[] = [
    'BasicAuthentication',
    'OAuth2ClientCredentials',
    'OAuth2SAMLBearerAssertion',
    'OAuth2UserTokenExchange',
    'OAuth2JWTBearer',
    'PrincipalPropagation'
  ];
  return authTypesWithAuthorizationHeader.includes(destination.authentication);
}

function getCustomAuthHeaders(
  destination: Destination,
  customHeaders?: Record<string, any>
) {
  if (destination.authentication === 'PrincipalPropagation') {
    return pickIgnoreCase(customHeaders, 'SAP-Connectivity-Authentication');
  }
  return pickIgnoreCase(customHeaders, 'authorization');
}

export async function getAuthHeaders(
  destination: Destination,
  customHeaders?: Record<string, any>
): Promise<Record<string, string>> {
  const customAuthHeaders = getCustomAuthHeaders(destination, customHeaders);

  if (Object.keys(customAuthHeaders).length && needsAuthHeaders(destination)) {
    logger.warn(
      'Found custom authorization headers. The given destination also provides authorization headers. This might be unintended. The custom headers from the request config will be used.'
    );
  }

  const additionalDestinationAuthHeaders = getCustomAuthHeaders(
    destination,
    destination.headers
  );

  if (Object.keys(customAuthHeaders).length) {
    return customAuthHeaders;
  }
  if (Object.keys(additionalDestinationAuthHeaders).length) {
    return additionalDestinationAuthHeaders;
  }
  return buildAuthorizationHeaders(destination);
}

/**
 * @deprecated Since v1.20.0. Use [[buildAuthorizationHeaders]] instead.
 * Adds authorization headers for a given destination to existing headers.
 *
 * @param destination - A destination.
 * @param headers - The headers that should be added to.
 * @returns The provided headers with the new authorization headers.
 */
export function buildAndAddAuthorizationHeader(destination: Destination) {
  return async function (
    headers: Record<string, any>
  ): Promise<Record<string, string>> {
    return {
      ...headers,
      ...(await buildAuthorizationHeaders(destination))
    };
  };
}

function toAuthorizationHeader(authorization: string): Record<string, string> {
  return toSanitizedObject('authorization', authorization);
}

function headerFromTokens(
  authenticationType: AuthenticationType,
  authTokens?: DestinationAuthToken[] | null
): Record<string, string> {
  if (!authTokens || !authTokens.length) {
    throw Error(
      `\`AuthenticationType\` is "${authenticationType}", but no auth tokens could be fetched from the destination service.`
    );
  }
  const usableTokens = authTokens.filter(
    (token: DestinationAuthToken) => !token.error
  );

  if (!usableTokens.length) {
    throw Error(
      [
        'The destination tried to provide authorization tokens but failed in all cases. This is most likely due to misconfiguration.',
        'Original error messages:',
        ...authTokens.map(token => token.error)
      ].join(unixEOL)
    );
  }
  const authToken = usableTokens[0];
  return toAuthorizationHeader(authToken.http_header.value);
}

function headerFromBasicAuthDestination(
  destination: Destination
): Record<string, string> {
  if (isNullish(destination.username) || isNullish(destination.password)) {
    throw Error(
      'AuthenticationType is "BasicAuthentication", but "username" and / or "password" are missing!'
    );
  }

  return toAuthorizationHeader(
    basicHeader(destination.username, destination.password)
  );
}

export function basicHeader(username: string, password: string): string {
  return 'Basic ' + encodeBase64(`${username}:${password}`);
}

function headerForPrincipalPropagation(
  destination: Destination
): Record<string, any> {
  const principalPropagationHeader =
    destination?.proxyConfiguration?.headers?.[
      'SAP-Connectivity-Authentication'
    ];
  if (!principalPropagationHeader) {
    throw Error(
      'Principal propagation was selected in destination, but no SAP-Connectivity-Authentication bearer header was added by connectivity service.'
    );
  }
  return {
    'SAP-Connectivity-Authentication': principalPropagationHeader
  };
}

function headerForProxy(destination: Destination): Record<string, any> {
  const proxyAuthHeader =
    destination?.proxyConfiguration?.headers?.['Proxy-Authorization'];

  return proxyAuthHeader ? { 'Proxy-Authorization': proxyAuthHeader } : {};
}

// TODO the proxy header are for OnPrem auth and are now handled correctly and should be removed here
// However this would be a breaking change, since we recommended to use 'NoAuthentication' to achieve principal propagation as a workaround.
// Remove this in v2
function legacyNoAuthOnPremiseProxy(
  destination: Destination
): Record<string, any> {
  logger.warn(
    `You are using 'NoAuthentication' in destination: ${destination.name} which is an OnPremise destination. This is a deprecated configuration, most likely you wanted to set-up 'PrincipalPropagation' so please change the destination property to the desired authentication scheme.`
  );

  let principalPropagationHeader;
  try {
    principalPropagationHeader = headerForPrincipalPropagation(destination);
  } catch (e) {
    logger.warn('No principal propagation header found.');
  }

  return {
    ...headerForProxy(destination),
    ...principalPropagationHeader
  };
}

function getProxyRelatedAuthHeaders(
  destination: Destination
): Record<string, any> {
  if (
    destination.proxyType === 'OnPremise' &&
    destination.authentication === 'NoAuthentication'
  ) {
    return legacyNoAuthOnPremiseProxy(destination);
  }

  // The connectivity service will raise an exception if it can not obtain the 'Proxy-Authorization' and the destination lookup will fail early
  return headerForProxy(destination);
}

async function getAuthenticationRelatedHeaders(
  destination: Destination
): Promise<Record<string, string>> {
  logger.debug(
    `Getting authentication related headers for authentication type: ${destination.authentication}`
  );

  switch (destination.authentication) {
    case null:
    case undefined:
      logger.warn(
        'No authentication type is specified on the destination! Assuming "NoAuthentication".'
      );
      return {};
    case 'NoAuthentication':
    case 'ClientCertificateAuthentication':
      return {};
    case 'OAuth2SAMLBearerAssertion':
    case 'OAuth2UserTokenExchange':
    case 'OAuth2JWTBearer':
    case 'OAuth2ClientCredentials':
      return headerFromTokens(
        destination.authentication,
        destination.authTokens
      );
    case 'BasicAuthentication':
      return headerFromBasicAuthDestination(destination);
    case 'PrincipalPropagation':
      return headerForPrincipalPropagation(destination);
    default:
      throw Error(
        `The destination used "${destination.authentication}" as authentication type which is not supported by the SAP Cloud SDK.`
      );
  }
}

export async function buildAuthorizationHeaders(
  destination: Destination
): Promise<Record<string, string>> {
  const sanitizedDestination = sanitizeDestination(destination);
  return {
    ...(await getAuthenticationRelatedHeaders(sanitizedDestination)),
    ...getProxyRelatedAuthHeaders(sanitizedDestination)
  };
}
