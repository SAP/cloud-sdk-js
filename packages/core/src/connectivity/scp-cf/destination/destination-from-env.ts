import { createLogger, ErrorWithCause } from '@sap-cloud-sdk/util';
import {
  proxyStrategy,
  ProxyStrategy,
  addProxyConfigurationInternet
} from '../../../http-agent/proxy-util';
import { decodeJwt } from '../jwt';
import {
  sanitizeDestination,
  isDestinationConfiguration,
  parseDestination
} from './destination';
import type {
  Destination,
  DestinationAuthToken
} from './destination-service-types';
import type { DestinationOptions } from './destination-accessor';

const logger = createLogger({
  package: 'core',
  messageContext: 'env-destination-accessor'
});

/**
 * Get all destinations from the environment variable "destinations".
 * This is discouraged for productive use! Use [[useOrFetchDestination]] for fetching destinations from the Cloud Foundry destination service.
 *
 * @returns A list of destinations
 */
export function getDestinationsFromEnv(): Destination[] {
  const destinationsEnv = getDestinationsEnvVariable();
  logger.debug(
    `The value for the destination environment variable is: ${destinationsEnv}`
  );
  if (destinationsEnv) {
    let destinations;
    try {
      destinations = JSON.parse(destinationsEnv);
    } catch (err) {
      throw new ErrorWithCause(
        'Error in parsing the destinations from the environment variable.',
        err
      );
    }
    validateDestinations(destinations);
    return destinations.map(destination =>
      isDestinationConfiguration(destination)
        ? parseDestination(destination)
        : sanitizeDestination(destination)
    );
  }
  return [];
}

/**
 * @deprecated Since v1.4.2. Use [[getDestinationsFromEnv]] instead.
 *
 * Get all destinations from the environment variable "destinations".
 * This is discouraged for productive use! Use destination-accessor/useOrFetchDestination for fetching destinations
 * from the Cloud Foundry destination service.
 *
 * @returns A list of destinations
 */
export function getDestinations(): Destination[] {
  return getDestinationsFromEnv();
}

/**
 * Get a destination from the environment variables by name. If there are multiple destinations with the same name the first one will be used.
 * This is discouraged for productive use! Use destination-accessor/useOrFetchDestination for fetching destinations
 * from the Cloud Foundry destination service.
 * @param name - Name of the destination
 * @returns The requested destination if existent, otherwise `null`
 */
export function getDestinationFromEnvByName(name: string): Destination | null {
  const matchingDestinations = getDestinationsFromEnv().filter(
    dest => dest.name === name
  );
  if (!matchingDestinations.length) {
    return null;
  }
  if (matchingDestinations.length > 1) {
    logger.warn(
      `The 'destinations' env variable contains multiple destinations with the name '${name}'. Only the first entry will be respected.`
    );
  }
  const destination = matchingDestinations[0];
  return proxyStrategy(destination) === ProxyStrategy.INTERNET_PROXY
    ? addProxyConfigurationInternet(destination)
    : destination;
}

/**
 * @deprecated Since v1.4.2. Use [[getDestinationFromEnvByName]] instead.
 *
 * Get a destination from the environment variables by name. Throws an error if there are multiple destinations with the same name.
 * This is discouraged for productive use! Use destination-accessor/useOrFetchDestination for fetching destinations
 * from the Cloud Foundry destination service.
 * @param name - Name of the destination
 * @returns The requested destination if existent, otherwise `null`
 */
export function getDestinationByName(name: string): Destination | null {
  return getDestinationFromEnvByName(name);
}

/* eslint-disable valid-jsdoc */

/**
 * @hidden
 */
export function getDestinationConfig(
  dest: string | Destination = 'ErpQueryEndpoint'
): Destination | null {
  return typeof dest === 'string' ? getDestinationFromEnvByName(dest) : dest;
}

/**
 * @hidden
 */
export function getDestinationsEnvVariable(): string | undefined {
  return process.env['destinations'];
}

function validateDestinations(destinations: any[]) {
  destinations.forEach(destination => {
    if (
      typeof destination.name === 'undefined' &&
      typeof destination.Name === 'undefined'
    ) {
      logger.warn(
        "Destination from 'destinations' env variable is missing 'name' or 'Name' property."
      );
    }
  });
}

/**
 * @hidden
 */
export function searchEnvVariablesForDestination(
  name: string,
  options: DestinationOptions = {}
): Destination | undefined {
  logger.info('Attempting to retrieve destination from environment variable.');

  if (getDestinationsEnvVariable()) {
    try {
      const destination = getDestinationFromEnvByName(name);
      if (destination) {
        if (destination.forwardAuthToken) {
          destination.authTokens = destinationAuthToken(options.userJwt);
          logger.info(
            `Successfully retrieved destination '${name}' from environment variable.`
          );
        } else {
          logger.warn(
            `Successfully retrieved destination '${name}' from environment variable.` +
              'This is discouraged for productive applications. ' +
              'Unset the variable to read destinations from the destination service on SAP Business Technology Platform.'
          );
        }
        return destination;
      }
    } catch (error) {
      logger.error(
        `Error in reading the given destinations from the environment variable ${error.message}.`
      );
    }
  }

  logger.info('No environment variable set.');
}

function destinationAuthToken(
  token?: string
): [DestinationAuthToken] | undefined {
  if (token) {
    const decoded = decodeJwt(token);
    logger.info(
      "Option 'forwardAuthToken' enabled on destination. Using the initial token for the destination."
    );
    return [
      {
        value: token,
        expiresIn: decoded.exp!.toString(),
        error: null,
        http_header: { key: 'Authorization', value: `Bearer ${token}` },
        type: 'Bearer'
      }
    ];
  }
  logger.warn(
    "Option 'forwardAuthToken' was set on destination but no token was provided to forward. This is most likely unintended and will lead to a authorization error on request execution."
  );
}
