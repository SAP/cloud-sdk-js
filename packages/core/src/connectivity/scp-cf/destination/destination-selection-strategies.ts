import { createLogger } from '@sap-cloud-sdk/util';
import type {
  AllDestinations,
  DestinationsByType
} from './destination-accessor-types';
import type { Destination } from './destination-service-types';

const logger = createLogger({
  package: 'core',
  messageContext: 'destination-selection-strategies'
});

export type DestinationSelectionStrategy = (
  allDestinations: AllDestinations,
  destinationName: string
) => Destination | null;

/**
 * Constraints the selection to provider destinations.
 *
 * @param allDestinations - Retrieved destinations
 * @param destinationName - Name of the destination to retrieve
 * @returns the destination to retrieve, returns null if no matched provider destination is found
 */
export function alwaysProvider(
  allDestinations: AllDestinations,
  destinationName: string
): Destination | null {
  return findDestination(allDestinations.provider, destinationName) || null;
}

/**
 * Constraints the selection to subscriber destinations.
 *
 * @param allDestinations - Retrieved destinations
 * @param destinationName - Name of the destination to retrieve
 * @returns the destination to retrieve, returns null if no matched subscriber destination is found
 */
export function alwaysSubscriber(
  allDestinations: AllDestinations,
  destinationName: string
): Destination | null {
  return findDestination(allDestinations.subscriber, destinationName) || null;
}

/**
 * Prioritizes the selection of subscriber destinations.
 *
 * @param allDestinations - Retrieved destinations
 * @param destinationName - Name of the destination to retrieve
 * @returns the destination to retrieve, returns null if no matched destination is found
 */
export function subscriberFirst(
  allDestinations: AllDestinations,
  destinationName: string
): Destination | null {
  return (
    findDestination(allDestinations.subscriber, destinationName) ||
    findDestination(allDestinations.provider, destinationName) ||
    null
  );
}

/**
 * Selector of destination selection strategies. See [[alwaysProvider]], [[alwaysSubscriber]] and [[subscriberFirst]] for more information available selection strategies.
 */
export const DestinationSelectionStrategies = {
  alwaysProvider,
  alwaysSubscriber,
  subscriberFirst
};

function findDestination(
  destinations: DestinationsByType,
  destinationName: string
): Destination | undefined {
  const isRequestedDestination = (destination: Destination) =>
    destination.name === destinationName;
  const instanceDest = destinations.instance.find(isRequestedDestination);
  const subAccountDest = destinations.subaccount.find(isRequestedDestination);

  if (instanceDest && subAccountDest) {
    logger.warn(
      `A destination with name ${destinationName} has been found for the destination serivce instance and subaccount. The instance destination will be used.`
    );
  }
  return instanceDest || subAccountDest;
}
