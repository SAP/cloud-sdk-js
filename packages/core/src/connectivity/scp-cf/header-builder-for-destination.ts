import { mergeLeftIgnoreCase, pickNonNullish } from '@sap-cloud-sdk/util';
import { Destination } from '../scp-cf';
import { getAuthHeaders } from './authorization-header';
export async function buildHeadersForDestination(
  destination: Destination,
  customHeaders?: Record<string, any>
): Promise<Record<string, string>> {
  const authHeaders = await getAuthHeaders(destination, customHeaders);

  const sapHeaders = mergeLeftIgnoreCase(
    pickNonNullish({
      'sap-client': destination.sapClient,
      'SAP-Connectivity-SCC-Location_ID': destination.cloudConnectorLocationId
    }),
    customHeaders
  );

  return { ...authHeaders, ...sapHeaders };
}
