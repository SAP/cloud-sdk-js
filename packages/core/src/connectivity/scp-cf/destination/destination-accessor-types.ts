import { Destination } from './destination-service-types';
import { DestinationSelectionStrategy } from './destination-selection-strategies';

/**
 * @hidden
 */
export interface AllDestinations {
  subscriber: DestinationsByType;
  provider: DestinationsByType;
}

/**
 * @hidden
 */
export interface DestinationsByType {
  instance: Destination[];
  subaccount: Destination[];
}

export interface DestinationAccessorOptions {
  /**
   * Method that implements the selection strategy of the retrieved destination. Uses [[subscriberFirst]] per default. Use the selector helper [[DestinationSelectionStrategies]] to select the appropriate selection strategy.
   */
  selectionStrategy?: DestinationSelectionStrategy;

  /**
   * The user token of the current request.
   */
  userJwt?: string;

  /**
   * @hidden
   */
  iss?: string;
  // FIXME This is used to put a subscriber domain in without having a JWT like for background processes.
  // We will create a seperate method for this on the destination accessor wit proper JS doc. This will be deprecated.
}
