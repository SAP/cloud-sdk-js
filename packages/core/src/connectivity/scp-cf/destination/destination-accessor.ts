import { VerifyJwtOptions } from '../jwt';
import { sanitizeDestination } from './destination';
import {
  Destination,
  DestinationNameAndJwt,
  DestinationRetrievalOptions,
  isDestinationNameAndJwt
} from './destination-service-types';
import { searchEnvVariablesForDestination } from './destination-from-env';
import { searchServiceBindingForDestination } from './destination-from-vcap';
import { getDestinationFromDestinationService } from './destination-from-service';
import { DestinationAccessorOptions } from './destination-accessor-types';

/**
 * Returns the parameter if it is a destination, calls [[getDestination]] otherwise (which will try to fetch the destination
 * from the Cloud Foundry destination service).
 *
 * Fetching a destination requires:
 * - a binding to exactly one XSUAA service instance with service plan "application"
 * - a binding to a destination service instance
 *
 * If either of the prerequisites is not met or one of the services returns an error, this function will either throw an error or return a promise that rejects.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @param options - Caching options by fetching destination.
 * @returns A promise resolving to the requested destination on success.
 */
export async function useOrFetchDestination(
  destination: Destination | DestinationNameAndJwt,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return isDestinationNameAndJwt(destination)
    ? getDestination(destination.destinationName, {
        userJwt: destination.jwt,
        ...options
      })
    : sanitizeDestination(destination);
}

export type DestinationOptions = DestinationAccessorOptions &
  DestinationRetrievalOptions &
  VerifyJwtOptions;

/**
 * @deprecated Since v1.0.1. Use [[getDestination]] instead.
 *
 * Retrieves a destination with the given name from the Cloud Foundry destination service.
 * Returns `null`, if no destination can be found.
 * Requires the following service bindings: destination, XSUAA
 * By default, selects subscriber over provider and instance over subaccount destinations.
 *
 * If the destinations are read from the environment, the jwt will be ignored.
 * @param name - The name of the destination to be retrieved.
 * @param options - The options of the fetching query of the destination that include the JWT of the current request and the strategy for selecting a destination.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestinationOptions(
  name: string,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return getDestination(name, options);
}

/**
 * Builds a destination from one of three sources (in the given order):
 * - from the environment variable "destinations"
 * - from service bindings
 * - from the destination service
 *
 * If you want to get a destination only from a specific source, use the corresponding function directly
 *  (`getDestinationFromEnvByName`, `destinationForServiceBinding`, `getDestinationFromDestinationService`).
 * @param name - The name of the destination to be retrieved.
 * @param options - Configuration for how to retrieve destinations from the destination service.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestination(
  name: string,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return (
    searchEnvVariablesForDestination(name) ||
    searchServiceBindingForDestination(name) ||
    getDestinationFromDestinationService(name, options)
  );
}
