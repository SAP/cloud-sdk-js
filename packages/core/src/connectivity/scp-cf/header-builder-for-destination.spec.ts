import { createLogger } from '@sap-cloud-sdk/util';
import { Destination } from './destination';
import { buildHeadersForDestination } from './header-builder-for-destination';
import { checkHeaders } from './authorization-header.spec';

describe('header builder for destination', () => {
  it('warns if the destination and request config contains authorization information', () => {
    const destination: Destination = {
      url: '',
      authentication: 'BasicAuthentication',
      username: 'SomeUser',
      password: 'SomePassword'
    };

    const logger = createLogger({
      package: 'core',
      messageContext: 'authorization-header'
    });
    const warnSpy = jest.spyOn(logger, 'warn');
    buildHeadersForDestination(destination, { authorization: 'SomeThing' });
    expect(warnSpy).toBeCalledWith(
      'Found custom authorization headers. The given destination also provides authorization headers. This might be unintended. The custom headers from the request config will be used.'
    );
  });

  it("should still add header if the old 'NoAuthorization' workaround is used", async () => {
    const destination = {
      url: '',
      authentication: 'NoAuthentication',
      proxyType: 'OnPremise',
      proxyConfiguration: {
        headers: {
          'SAP-Connectivity-Authentication': 'someValue',
          'Proxy-Authorization': 'someProxyValue'
        }
      }
    } as Destination;

    const headers = await buildHeadersForDestination(destination);
    checkHeaders(headers);
  });

  it('adds location id headers if there is a cloudConnectorLocationId in the destination', async () => {
    const destination = {
      url: 'url',
      cloudConnectorLocationId: 'locationId'
    } as Destination;

    const headers = await buildHeadersForDestination(destination);
    expect(headers['SAP-Connectivity-SCC-Location_ID']).toEqual('locationId');
  });

  it('custom SAP headers take precedence over SDK built headers', async () => {
    const destination = {
      url: 'url',
      sapClient: 'destinationProperty'
    } as Destination;

    const headers = await buildHeadersForDestination(destination, {
      'sap-client': 'customHeader'
    });
    expect(headers['sap-client']).toEqual('customHeader');
  });

  it('destination SAP headers take precedence over SDK built headers', async () => {
    const destination = {
      url: 'url',
      sapClient: 'destinationProperty',
      headers: {
        'sap-client': 'destinationHeader'
      }
    } as Destination;

    const headers = await buildHeadersForDestination(destination);
    expect(headers['sap-client']).toEqual('destinationHeader');
  });

  it('custom SAP headers take precedence over destination headers', async () => {
    const destination = {
      url: 'url',
      sapClient: 'destinationProperty',
      headers: {
        'sap-client': 'destinationHeader'
      }
    } as Destination;

    const headers = await buildHeadersForDestination(destination, {
      'sap-client': 'customHeader'
    });
    expect(headers['sap-client']).toEqual('customHeader');
  });

  it('includes destination headers', async () => {
    const destination = {
      url: 'url',
      sapClient: 'destinationProperty',
      headers: {
        'some-header': 'some header',
        'header-to-overwrite': 'destinationHeader',
        'SAP-client': 'destinationHeader'
      }
    } as Destination;

    const headers = await buildHeadersForDestination(destination, {
      'sap-CLIENT': 'customHeader',
      'HEADER-to-overwrite': 'customHeader'
    });
    expect(headers).toEqual({
      'some-header': 'some header',
      'sap-CLIENT': 'customHeader',
      'HEADER-to-overwrite': 'customHeader'
    });
  });
});
