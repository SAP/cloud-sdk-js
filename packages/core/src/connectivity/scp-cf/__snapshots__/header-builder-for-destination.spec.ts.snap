// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[deprecated] does not throw on Principal Propagation 1`] = `"Principal propagation was selected in destination, but no SAP-Connectivity-Authentication bearer header was added by connectivity service."`;

exports[`buildAuthorizationHeaders OAuth2ClientCredentials should throw an error if tokenServiceUrl, client_id or client_secret are not set 1`] = `"Request for \\"OAuth2ClientCredentials\\" authentication access token failed or denied."`;

exports[`buildAuthorizationHeaders OAuth2ClientCredentials should throw an error when the access token to the request header for "OAuth2ClientCredentials" is rejected 1`] = `"Request for \\"OAuth2ClientCredentials\\" authentication access token failed or denied."`;

exports[`buildAuthorizationHeaders auth token errors Throws an error if no authTokens are present, password is null and authenticationType is wrongly set 1`] = `"AuthenticationType is \\"BasicAuthentication\\", but \\"username\\" and / or \\"password\\" are missing!"`;

exports[`buildAuthorizationHeaders auth token errors Throws an error if no authTokens are present, username is null and authenticationType is wrongly set 1`] = `"AuthenticationType is \\"BasicAuthentication\\", but \\"username\\" and / or \\"password\\" are missing!"`;

exports[`buildAuthorizationHeaders auth token errors throws an error if the error property is truthy for all provided authTokens 1`] = `
"The destination tried to provide authorization tokens but failed in all cases. This is most likely due to misconfiguration.
Original error messages:
error
error"
`;

exports[`buildAuthorizationHeaders principal propagation does not throw on Principal Propagation 1`] = `"Principal propagation was selected in destination, but no SAP-Connectivity-Authentication bearer header was added by connectivity service."`;
